slatedb/src/garbage_collector/compacted_gc.rs

```diff
@@ -0,0 +1,102 @@
+use crate::{
+    config::GarbageCollectorDirectoryOptions, db_state::SsTableId, manifest::store::ManifestStore,
+    tablestore::TableStore, SlateDBError,
+};
+use chrono::{DateTime, Utc};
+use std::collections::HashSet;
+use std::{sync::Arc, time::Duration};
+use tokio::time::Interval;
+use tracing::error;
+
+use super::{GcStats, GcTask, DEFAULT_INTERVAL, DEFAULT_MIN_AGE};
+
+pub(crate) struct CompactedGcTask {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    stats: Arc<GcStats>,
+    compacted_options: Option<GarbageCollectorDirectoryOptions>,
+}
+
+impl CompactedGcTask {
+    pub fn new(
+        manifest_store: Arc<ManifestStore>,
+        table_store: Arc<TableStore>,
+        stats: Arc<GcStats>,
+        compacted_options: Option<GarbageCollectorDirectoryOptions>,
+    ) -> Self {
+        CompactedGcTask {
+            manifest_store,
+            table_store,
+            stats,
+            compacted_options,
+        }
+    }
+
+    fn compacted_sst_min_age(&self) -> chrono::Duration {
+        let min_age = self
+            .compacted_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        chrono::Duration::from_std(min_age).expect("invalid duration")
+    }
+
+    async fn list_active_l0_and_compacted_ssts(&self) -> Result<HashSet<SsTableId>, SlateDBError> {
+        let active_manifests = self.manifest_store.read_active_manifests().await?;
+        let mut active_ssts = HashSet::new();
+        for manifest in active_manifests.values() {
+            for sr in manifest.core.compacted.iter() {
+                for sst in sr.ssts.iter() {
+                    active_ssts.insert(sst.id);
+                }
+            }
+            for sst in manifest.core.l0.iter() {
+                active_ssts.insert(sst.id);
+            }
+        }
+        Ok(active_ssts)
+    }
+}
+
+impl GcTask for CompactedGcTask {
+    /// Collect garbage from the compacted SSTs. This will delete any compacted SSTs that are
+    /// older than the minimum age specified in the options and are not active in the manifest.
+    async fn collect(&self, utc_now: DateTime<Utc>) -> Result<(), SlateDBError> {
```

Comment:
does this `collect()` method also have plans to have cancellation token handled? Its execution might also be relatively time consuming imo.

(just for discussion, if needed, I think it will be nice to have a separate PR for it.)

Reply:
I didn't think about it, but it might be a good idea. I just copied the three gc tasks from the root of the module to make the code easier to understand.

Reply:
i love this separation a lot 👍, each file has a single responsibility, which makes it much easier for readers to understand what they do